[quant][ns] Fix ns tool bug for mobilenetv2/v3

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74149

Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

terrychenism committed:
[quant][ns] Fix ns tool bug for mobilenetv2/v3

Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

vkuzo commented:
thanks for fixing this, would it be possible to add a reproducible test plan?

jerryzh168 commented on test/quantization/eager/test_numeric_suite_eager.py:
```diff
@@ -421,14 +421,14 @@ def test_compare_model_outputs_functional_static(self):
             "myadd_relu.stats",
-            "my_scalar_add.stats",
-            "my_scalar_mul.stats",
+            # "my_scalar_add.stats", # empty
```
is this change relevant?

vkuzo replied:
+1, what's the context for this?

terrychenism replied:
this tensor is empty, it would be filtered after this PR.

vkuzo replied:
in that case, I think it would be better to delete the lines instead of commenting them out

terrychenism replied:
sure

jerryzh168 commented on torch/ao/ns/_numeric_suite.py:
```diff
@@ -436,6 +436,8 @@ def get_matching_activations(
     quantized_dict = get_logger_dict(q_module)
     act_dict: Dict[str, Dict] = {}
     for key in quantized_dict:
+        if len(quantized_dict[key]["tensor_val"]) == 0:
```
do we need to add a test for this change?

vkuzo reviewed (changes requested):
requesting changes for test plan

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
due to the empty the tensor, fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

vkuzo commented on test/quantization/eager/test_numeric_suite_eager.py:
```diff
@@ -534,3 +537,46 @@ def test_shadow_logger(self):
+    @skip_if_no_torchvision
+    def test_mobilenet(self):
+        from torchvision.models.quantization import mobilenet_v2, mobilenet_v3_large
+        self._test_vision_model(mobilenet_v2(pretrained=True, quantize=False))
```
can we do separate test cases for these two models, instead of combining into one test case?

terrychenism replied:
updated.

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

vkuzo reviewed (approved):
looks good, thanks for fixing this!

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).

terrychenism committed:
Update on "[quant][ns] Fix ns tool bug for mobilenetv2/v3"


Summary:
mobilenet v2/v3 failed when using ns tool to analysis the model
original ns tool would generate the empty tensor, they cannot pass the function of error calculation. we fixed it by filtering the empty tensor

Test Plan:
python3 test/test_quantization.py TestNumericSuiteEager.test_mobilenet

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34851886](https://our.internmc.facebook.com/intern/diff/D34851886)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851886).
